### PR TITLE
maintenance(www): Explicitly type GraphQL fields

### PR DESCRIPTION
**Pull request by tesseralis**
## Description

Define explicit GraphQL types for sources in `www`. This lets us run the site without sourcing `docs/`.

* Create typings for `MdxFrontmatter`, `SitesYaml` and other fields used in the gatsby site.
* Add `/* GraphQL */` directive so that GraphQL is syntax-highlighted.

## Related Issues

#24913

### CODE REVIEW

File: www/src/utils/node/showcase.js [hunk 1]
```diff
@@ -2,12 +2,40 @@ const slugify = require(`slugify`)
...
+    type SitesYamlFields @dontInfer {
+      slug: String!
+      hasScreenshot: Boolean
+    }
+
+    # TODO this should be in gatsby-transformer-screenshot
+    type Screenshot implements Node @dontInfer {
+      screenshotFile: File
+    }
+  `)
```
Comment (tesseralis) [inline www/src/utils/node/showcase.js:30]: This is unneeded if #24944 goes in (and the dependency is updated)

File: docs/blog/2018-11-07-gatsby-for-apps/index.md [hunk 2]
```diff
@@ -5,7 +5,7 @@ author: Dustin Schau
 tags:
   - apps
 excerpt: Gatsby is great for not only static sites but also traditional web applications. Using Gatsby enables the benefits of both static and web applications so you don't have to sacrifice the advantages of one approach to reap the benefits of the other.
-cover: images/what-if-i-told-you.jpg
+image: images/what-if-i-told-you.jpg
```
Comment (tesseralis) [inline docs/blog/2018-11-07-gatsby-for-apps/index.md:8]:
```suggestion
cover: images/what-if-i-told-you.jpg
```

**Comment by tesseralis** [comment]
I'm getting the following errors on build:

```
warn Deprecation warning - adding inferred resolver for field Frontmatter.image. In Gatsby v3, only fields with an explicit
warn Deprecation warning - adding inferred resolver for field Frontmatter.cover. In Gatsby v3, only fields with an explicit
warn The type `SitesYaml` does not explicitly define the field `childScreenshot`.
On types with the `@dontInfer` directive, or with the `infer` extension set to `false`, automatically adding fields for children types
 is deprecated.
In Gatsby v3, only children fields explicitly set with the `childOf` extension will be added.
warn The type `StartersYaml` does not explicitly define the field `childScreenshot`.
On types with the `@dontInfer` directive, or with the `infer` extension set to `false`, automatically adding fields for children types
 is deprecated.
In Gatsby v3, only children fields explicitly set with the `childOf` extension will be added.
warn There are conflicting field types in your data.
```

Which is leading to:

```
warn You can't use childImageSharp together with undefined.undefined — use publicURL instead. The childImageSharp portion of the query
 in this file will return null:
```

I'm trying to read up on this [blog post](https://www.gatsbyjs.org/blog/2019-05-17-improvements-to-schema-customization/) on schema customization to learn more but I'm not quite sure what the right solution on typing would be. @freiksenet do you have additional context on this?

**Comment by freiksenet** [comment]
This is awesome! Should we mark all types that we define as @dontInfer or do some of them cause problems?

**Comment by tesseralis** [comment]
@freiksenet the `@dontInfer` causes problems with images. For example, if you add the `@dontInfer` onto `AuthorYaml`, it ends up with none of the `avatar` File fields loading correctly, causing this error:

```
warn You can't use childImageSharp together with undefined.undefined — use publicURL instead. The childImageSharp portion of the query
 in this file will return null:
```

If you add `@dontInfer` to `SitesYaml` or `StartersYaml`, it adds the following warning for the `childScreenshot` component:

```
warn The type `SitesYaml` does not explicitly define the field `childScreenshot`.
On types with the `@dontInfer` directive, or with the `infer` extension set to `false`, automatically adding fields for children types
 is deprecated.
In Gatsby v3, only children fields explicitly set with the `childOf` extension will be added.
```

Do you know another way to resolve this? Ideally we would be able to add the `@dontInfer` so that people don't accidentally add new fields to types.

**Pull request merged**